File: 64-minimum-path-sum/minimum-path-sum.ts
function minPathSum(grid: number[][]): number {
    let r = grid.length;
    let c = grid[0].length;
    if (r === 1 && c === 1) {
        return grid[0][0];
    }
    const memo = Array.from({length:r},()=> Array(c).fill(-1));
    function getMinPathSum(i:number,j:number){
        if(i==0 && j==0){
            return grid[0][0];
        }
        if(i<0 || j<0){
            return Infinity
        }
        if(memo[i][j]!==-1){
            return memo[i][j]
        }
        let up = getMinPathSum(i-1,j);
        let left = getMinPathSum(i,j-1);
        return memo[i][j] = grid[i][j]+ Math.min(up,left);
    }
    getMinPathSum(r-1,c-1);
    return memo[r-1][c-1];
};

File: 64-minimum-path-sum/minimum-path-sum.py
class Solution(object):
    def minPathSum(self, grid):
        rows, cols = len(grid), len(grid[0])
        memo = [[-1 for _ in range(cols)] for _ in range(rows)]
        
        def minP(i, j):
            if i >= rows or j >= cols:
                return float('inf')
            
            if i == rows - 1 and j == cols - 1:
                return grid[i][j]
            
            if memo[i][j] != -1:
                return memo[i][j]
            
            memo[i][j] = grid[i][j] + min(minP(i, j + 1), minP(i + 1, j))
            return memo[i][j]
        
        
        return minP(0, 0)
